src/backend/distributed/commands/multi_copy.c

```diff
@@ -0,0 +1,1177 @@
+	PG_TRY();
+	{
+		while (true)
+		{
+			bool nextRowFound = false;
+			Datum partitionColumnValue = 0;
+			ShardInterval *shardInterval = NULL;
+			int64 shardId = 0;
+			bool found = false;
+			MemoryContext oldContext = NULL;
+
+			ResetPerTupleExprContext(executorState);
+
+			oldContext = MemoryContextSwitchTo(executorTupleContext);
+
+			/* parse a row from the input */
+			nextRowFound = NextCopyFrom(copyState, executorExpressionContext,
+										columnValues,columnNulls, NULL);
+
+			MemoryContextSwitchTo(oldContext);
+
+			if (!nextRowFound)
+			{
+				break;
+			}
+
+			CHECK_FOR_INTERRUPTS();
+
+			/* find the partition column value */
+
+			if (columnNulls[partitionColumn->varattno - 1])
+			{
+				ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+								errmsg("cannot copy row with NULL value "
+									   "in partition column")));
+			}
+
+			partitionColumnValue = columnValues[partitionColumn->varattno - 1];
+
+			/* find the shard interval and id for the partition column value */
+			shardInterval = FindShardInterval(partitionColumnValue, shardIntervalCache,
+											  shardCount, partitionMethod,
+											  compareFunction, hashFunction,
+											  useBinarySearch);
+			if (shardInterval == NULL)
+			{
+				ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+								errmsg("no shard for partition column value")));
+			}
+
+			shardId = shardInterval->shardId;
+
+			/* find the connections to the shard placements */
+			shardConnections = (ShardConnections *) hash_search(shardConnectionHash,
+																&shardInterval->shardId,
+																HASH_ENTER,
+																&found);
+			if (!found)
+			{
+				/* intialize COPY transactions on shard placements */
+				shardConnections->shardId = shardId;
+				shardConnections->connectionList = NIL;
+
+				OpenCopyTransactions(copyStatement, shardConnections, shardId);
+
+				BuildCopyBinaryHeaders(copyOutState);
+				SendCopyDataToPlacements(copyOutState->fe_msgbuf, shardConnections);
+			}
+
+			/* Replicate row to all shard placements */
+			BuildCopyRowData(columnValues, columnNulls, tupleDescriptor,
+							 copyOutState, columnOutputFunctions);
+			SendCopyDataToPlacements(copyOutState->fe_msgbuf, shardConnections);
+
+			processedRowCount += 1;
+		}
+
+		/* send binary footers to all shards */
+		hash_seq_init(&status, shardConnectionHash);
+
+		shardConnections = (ShardConnections *) hash_seq_search(&status);
+		while (shardConnections != NULL)
+		{
+			BuildCopyBinaryFooters(copyOutState);
+			SendCopyDataToPlacements(copyOutState->fe_msgbuf, shardConnections);
+
+			shardConnections = (ShardConnections *) hash_seq_search(&status);
+		}
+
+		connectionList = ConnectionList(shardConnectionHash);
+
+		EndRemoteCopy(connectionList, true);
+
+		if (CopyTransactionManager == TRANSACTION_MANAGER_2PC)
+		{
+			PrepareTransactions(connectionList);
+		}
+
+		pfree(columnValues);
```

Comment:
>
> No need to free, the memory context this is run under will be reset shortly after.
>
> I was thinking the same, but looking into CopyFrom() in copy.c, it keeps
pfree() calls. Is something different there?

For code clearity, it seems better to delete them. I will also delete
FreeExecutorState(executorState) at the same place.